### ISSUE TITLE
nested .ungroup_stats in make_afun

### ISSUE THREAD

**Issue by anajens** [Roche/rtables#240]
Hi @gmbecker ,

I'm wondering if `make_afun` can work with nested statistics and preserve the label groups? Here's a simple example:

```
> s_grp <- function(df, a = 1, b = 2) {
+   
+   list(
+     letters = list(
+       A = list(a1 = a, a2 = a*2),
+       B = list(b1 = b, b2 = b*2, b3 = b*3)
+     )
+   )
+   
+ }
> 
> s_grp(iris)
$letters
$letters$A
$letters$A$a1
[1] 1

$letters$A$a2
[1] 2


$letters$B
$letters$B$b1
[1] 2

$letters$B$b2
[1] 4

$letters$B$b3
[1] 6

> a_grp <- make_afun(s_grp, .ungroup_stats ="letters")
> 
> a_grp(iris)
RowsVerticalSection (in_rows) object print method:
----------------------------
  row_name formatted_cell indent_mod row_label
1        A           1, 2          0         A
2        B        2, 4, 6          0         B
```

Ideal output in the layout at would look like this (this is without any split functions being applied):
```
A
  a1 1
  a2 2
B
  b1 2
  b2 4
  b3 6
```

**Comment by gmbecker** [comment]
Hi Jana,

So generally an analysis function shouldn't cause that appearance by
itself, because then the appearance radically differs from the structure
(ie, there would be no c(*, "A", *, "a1") path within the table, even
though the output makes it really look like there should be.  This seems
like it would be really painful and confusing for the users.

That said, its likely you could manually write an afun that forced this
behavior if you really wanted to (though A and B would not be label rows,
they'd be data rows with "" in all their cells.

Can you elaborate a bit more on why you would want the appearance of a row
split without performing a row split?

~G

**Comment by anajens** [comment]
Hi Gabe, thanks for the clarification. I'm also reluctant to create a layout that differs from the underlying data structure.

Why this is needed is because the 2 groups under "A" and "B" share a denominator. So the full layout looks something like this:

```
Grp (n)     xx <- this is from summarize_row_groups()
  A
    a1     1 / xx
    a2     2 / xx
  B
    b1     2 / xx
    b2     4 / xx
    b3     6 / xx
```

So I think the solution you had in mind for #117 should be able to solve this because then we could introduce a split on the variable containing `A` and `B` and still get the right denominator. Do you agree?

I think in the meantime our best alternative to try is just to produce a layout like this which omits the "label" rows. The downsides are that the numbers would be correct but it's a bit more difficult to read and deviates from our standard template.
```
Grp (n)     xx 
    a1     1 / xx
    a2     2 / xx
    b1     2 / xx
    b2     4 / xx
    b3     6 / xx
```

**Comment by gmbecker** [comment]
@anajens @danielinteractive So this is now covered by the `.spl_context` argument split, content, and analysis functions can all take. when your functions accept`.spl_context`, they will receive a data.frame with `split`, `value`, and `full_parent_df` columns. 

If they are in a row context (ie row splitting, content functions, or analysis functions), the data.frame will additional have a column for each column in the table structure (named the same thing as `names(col_exprs(tab))`), containing logical expressiosn which represent subsetting `full_parent_df`  for that column. Finally, for convenience, they will have `cur_col_subset` and `cur_col_n`  which correspond to the column currently being operated in.

`?split_cols_by`, `?split_rows_by`, `?analyze` all now reflect this. 

See the following example code from the tests I added. Here the content row for STRATA1  has the country total observation (ie row grouping) count in its label (in parentheses) and then has counts per column for the strata out of counts per column for the country in its cells. 

The analysis function counts number of observations where the age is greater than the global mean, but uses the STRATA1 column counts as its 'denominator'. I believe this should cover nearly all the outstanding row denominator use cases *other than reference row grouping support*. :

```
 ageglobmean <- mean(DM$AGE)
    cfun <- function(df, labelstr, .spl_context) {
            lastrow <- .spl_context[nrow(.spl_context) - 1,]
            in_rows(c(nrow(df), lastrow$cur_col_n),
                    .names = labelstr,
                    .labels = sprintf("%s (%d)", labelstr, nrow(lastrow$full_parent_df[[1]])),
                    .formats = "xx / xx")
    }

    afun <- function(x, .spl_context) {
        lastrow <- .spl_context[nrow(.spl_context),]
        in_rows(c(sum(x >= ageglobmean), lastrow$cur_col_n),
                .names = "age_analysis",
                .labels = sprintf("counts (out of %d)", nrow(lastrow$full_parent_df[[1]])),
                .formats = "xx / xx")
    }


    lyt <- basic_table() %>%
        split_cols_by("ARM") %>%
        split_cols_by("SEX", split_fun = keep_split_levels(c("M", "F"))) %>%
        split_rows_by("COUNTRY", split_fun = keep_split_levels(c("CHN", "USA"))) %>%
        summarize_row_groups() %>%
        split_rows_by("STRATA1") %>%
        summarize_row_groups(cfun = cfun) %>%
        analyze("AGE", afun = afun)

    tab <-  build_table(lyt, DM)
```
```
                                A: Drug X               B: Placebo             C: Combination     
                             M            F           M           F            M            F     
--------------------------------------------------------------------------------------------------
CHN                      28 (54.9%)   34 (48.6%)   19 (38%)   29 (51.8%)   38 (55.9%)   31 (50.8%)
  A (179)                  7 / 28      12 / 34      2 / 19     12 / 29      11 / 38       9 / 31  
    counts (out of 53)     4 / 7        3 / 12      0 / 2       5 / 12       3 / 11       5 / 9   
  B (179)                  8 / 28      11 / 34      7 / 19      9 / 29      12 / 38      10 / 31  
    counts (out of 57)     4 / 8        5 / 11      2 / 7       3 / 9        5 / 12       4 / 10  
  C (179)                 13 / 28      11 / 34     10 / 19      8 / 29      15 / 38      12 / 31  
    counts (out of 69)     9 / 13       6 / 11      3 / 10      7 / 8        5 / 15       6 / 12  
USA                       5 (9.8%)    8 (11.4%)    8 (16%)    6 (10.7%)    7 (10.3%)    10 (16.4%)
  A (44)                   1 / 5        4 / 8       1 / 8       2 / 6        4 / 7        5 / 10  
    counts (out of 17)     1 / 1        2 / 4       1 / 1       0 / 2        2 / 4        3 / 5   
  B (44)                   2 / 5        1 / 8       6 / 8       0 / 6        1 / 7        1 / 10  
    counts (out of 11)     2 / 2        1 / 1       2 / 6       0 / 0        0 / 1        1 / 1   
  C (44)                   2 / 5        3 / 8       1 / 8       4 / 6        2 / 7        4 / 10  
    counts (out of 16)     1 / 2        2 / 3       0 / 1       2 / 4        1 / 2        3 / 4  
```

**NB currently these column counts do not reflect overridden column counts, they are data split row counts. This is explicitly documented as the current behavior.**

**Comment by anajens** [comment]
Hi @gmbecker ,

when I look into `.spl_context` data.frame from example above for either the afun or cfun I see the following columns:

```
Browse[1]> names(.spl_context)
[1] "split"          "value"          "full_parent_df" "cur_col_subset" "cur_col_n"   
```

This doesn't match the docs:
![image](https://user-images.githubusercontent.com/43623360/134442173-0b7f33cc-e7df-4e7e-8170-ef30ea528093.png)

Am I missing something?

**Comment by gmbecker** [comment]
Ah, yes you are correct, I strip them out apparently when know I'm going to put the cur_col_* ones in.

Should be an eeasy fix.

**Comment by gmbecker** [comment]
Fixed in 0.3.8.9011 on gabe_tabletree_work branch @anajens

**Comment by anajens** [comment]
Thanks Gabe!

This is working really nicely with `make_afun` and below I show simplified example based on the initial question on this post. I think this issue can be closed.

```r
df <- data.frame(
  id = c(1:15, 1:10),
  param = c(rep("lab xxx", 15), rep("lab yyy", 10)),
  abn = factor(c(rep("low", 3), rep("high", 12), rep("low", 10)), levels = c("low", "high")),
  grade = c("1", rep("2", 2), rep("1", 2), rep("2", 4), rep("3", 6), rep("1", 7), rep("2", 2), "3")
)

# statistics function
sfun <- function(x, .spl_context) {
  
  first_row <- .spl_context[1,]

  x_lvls <- levels(x)
  
  result <- split(numeric(0), factor(x_lvls))
  
  for (lvl in x_lvls) {
    result[[lvl]] <- with_label(c(num = sum(x == lvl), denom = first_row$cur_col_n), lvl)
  }
  
  result <- list(fraction = result)
  result
}

# formatted afun
afun <- make_afun(sfun, .formats = c(fraction = "xx / xx"), .ungroup_stats = "fraction")

# Try it out
sfun(df$grade[df$param == "lab yyy"], .spl_context = data.frame(cur_col_n = 10, all_cols_n = 20))

afun(df$grade[df$param == "lab yyy"], .spl_context = data.frame(cur_col_n = 10, all_cols_n = 20))

basic_table() %>%
  split_rows_by("param") %>%
  summarize_row_groups() %>%
  split_rows_by("abn", split_fun = trim_levels_in_group("grade")) %>%
  analyze("grade", afun = afun) %>%
  build_table(df, col_counts = 100)
```

```
> sfun(df$grade[df$param == "lab yyy"], .spl_context = data.frame(cur_col_n = 10, all_cols_n = 20))
$fraction
$fraction$`1`
  num denom 
    7    10 
attr(,"label")
[1] "1"

$fraction$`2`
  num denom 
    2    10 
attr(,"label")
[1] "2"

$fraction$`3`
  num denom 
    1    10 
attr(,"label")
[1] "3"
```

```
> afun(df$grade[df$param == "lab yyy"], .spl_context = data.frame(cur_col_n = 10, all_cols_n = 20))
RowsVerticalSection (in_rows) object print method:
----------------------------
  row_name formatted_cell indent_mod row_label
1        1         7 / 10          0         1
2        2         2 / 10          0         2
3        3         1 / 10          0         3
```

Result:
```
          all obs 
          (N=100) 
------------------
lab xxx   15 (15%)
  low             
    1      1 / 15 
    2      2 / 15 
  high            
    1      2 / 15 
    2      4 / 15 
    3      6 / 15 
lab yyy   10 (10%)
  low             
    1      7 / 10 
    2      2 / 10 
    3      1 / 10 
```
